### ISSUE TITLE
Utoipa users poll: How do you use the responses and how would you like them to work? Answer would help.

### ISSUE THREAD

**Issue by juhaku** [juhaku/utoipa#1068]
## Responses poll

Currently while I am working on refactoring multiple aspects of the `utoipa` lib for the coming release 5.0.0 this is good place to introduce some improvements to current workflow within boundaries set by its users. So here comes the question.

How do you currently use the responses with `utoipa` lib? 
1. Only tuples
2. Tuples with  derived `ToResponse` 
3. Derived IntoResponses
4. Any other way, please describe. 

Second question is then how would you like the responses work, or is there any aspect you would like to change in order to them to be more user friendly?

**Comment by michaelvlach** [comment]
Hi, first of all huge thank you for this crate!

To answer your questions. I use dedicated structs and deriving from `ToSchema`: https://github.com/agnesoft/agdb/blob/main/agdb_api/rust/src/api_types.rs

I find it quite nice and I would not change it. There are two aspects I think would be nice if improved but I am not sure if possible:

1) I miss validation of the stuff in the derived against actual code (route signature / types). Obviously things like status codes cannot be validated but I hope types could be. Often I missed the change in return type in the derive annotation and nothing complained. Luckily I have comprehensive test suite so I caught it but a compile error would be nice if there was a mismatch. Probably not possible though.

2) More comprehensive documentation on the derive macro. I use a lots of stuff from utoipa and sometimes it was difficult to find out how to achieve what I wanted. E.g. https://github.com/agnesoft/agdb/blob/main/agdb_server/src/routes/db.rs#L82 Since I generate apis for different languages and they require things like tags, operation_id etc. That is a common struggle, no such thing as enough docs. :-)

**Comment by juhaku** [comment]
Great, thanks for a reply.

1. There is improvement for this actually coming in the `5.0.0` release of utoipa. However if you completely forget the add the `response(...)` declaration then the improvement will not help there yet since the return types are not inferred. But could be with certain limitations. In the `5.0.0` release the `request_body` and response `body` schemas will be correctly spanned thus it will give compile error if an item does not exists and will also get renamed if you rename the type as well.
2. True, more docs, that is forever target of improvement. :slightly_smiling_face:

**Comment by juhaku** [comment]
> Hey, not really an answer on how I uses it, but I would really like to be able to share responses between routes.

Responses can be shared if used with `ToResponse`. Types that implement `ToResponse` can be added to the responses section of OpenApi. and then they can be references with `response = MyResponse`.
```rust

struct MyResponse; 

impl<'s> ToResponse<'s> for MyResponse {
  fn response() -> (&'s str, Response) -> {
    // add all the examples here programmatically
    ("MyResponse", Response::builder().build())
  }
}

#[utoipa::path(
    post,
    path = "/{taskUid}",
    tag = "Tasks",
    params(("taskUid", format = UInt32, example = 0, description = "The task identifier", nullable = false)),
    responses(
        (status = 200, response = inline(MyResponse))
    )
)]
async fn get_task(} {}
```
You could also derive the `ToResponse` but I am most likely going to change the `ToResponse` implementation. And I am not sure about the current behavior of the derive. I think it lacking something but I am not sure what, and perhaps I am going to remove that completely if I find some better way to do shareable responses.

`example` and `examples` attributes could be allowed to use method references though nonetheless but hasn't been done so far.

**Comment by irevoire** [comment]
Humm, yes, I tried to use it, but it didn’t work, and I don’t really have the time to understand why, It's probably an issue on my side.

But in terms of API, I find it strange that the `ResponseBuilder` doesn’t contain the HTTP status code and the schema 🤔 
But it’s pretty nice to be able to define multiple examples for one type.
Defining and using the examples through a structure is also pretty good, IMO

Ideally, in our case, almost every route can have three different errors that share the same error code: If you forget to specify an API key, we don’t understand the format of your API key, or your API key doesn’t give you the right to access this route.
Iirc the first two returns an HTTP 400 while the last one return a 401 or something.
Then, they all share the same schema.
It would be really nice to be able to define all that on a single struct and then use it easily in most of my routes:
```rust
#[utoipa::path(
    post,
    path = "/{taskUid}",
    tag = "Tasks",
    params(("taskUid", format = UInt32, example = 0, description = "The task identifier", nullable = false)),
    responses(
        inline(ApiKeyError), // here I can define multiple examples with multiple HTTP status code
        inline(SummarizedTaskView), // here I only have one example
        (status = BAD_REQUEST, schema = ResponseError, example = json!(...)),
    )
)]
async fn get_task(} {}
```

I hope that can help you in your decisions!

**Comment by Aderinom** [comment]
Spend way too much time looking into this 😆 
What I'd care deeply about is the Good Path Response being as easy to create as possible.
<details>
  <summary><b>My reasoning/Viewpoint</b></summary>
For us usually following is true:

- Unless an error has technical implications and should   
    cause the caller to do anything but log, we usually don't bother documenting these.
- This e.g. on Bad Requests, Bad Auth, Internal Errors, Not Found etc.

Why:

1. Because IMO on most apis these things can and will not be Programatically handled by the caller,  
   just logged. As such spending time on perfect docs can be spend on something more effective.
2. Because each documented error asks the caller to think about how to handle this case specifically  
   so it adds mental fatigue and makes apis look much harder to implement than they actually are. 

However: There are some rare cases where I want percise control over the errors, I would be fine having a a litte more clunky api for these cases.

</details>
<br/>

At the moment, we usually do the tuple based approach with impl Responder as the return - so e.g.

```rs
    #[derive(Debug, serde::Serialize, serde::Deserialize, ToSchema)]
    pub struct Task {
        ...
    }

    /// Retrieves a certain task by id
    #[utoipa::path(
        responses(
            (status = 200, body = dto::Task),
        )
    )]
    #[get("/tasks/{id}")]
    pub async fn get_task_by_id(id : web::Path<i32>) -> impl Responder {
        println!("Should now get a task with id {}", id.into_inner());
        Ok(Json(Task:rand()))
    }
    

```

My dream would be: 

```rs

    #[derive(Debug, serde::Serialize, serde::Deserialize, ToSchema, IntoResponses)]
    #[response(status = default_ok)] // Maybe allow this to be optional ? No defined = default ok - So Get 200, Post 201...
    pub struct Task {
        ...
    }

    /// Retrieves a certain task by id
    #[utoipa::path]
    #[get("/tasks/{id}")]
    pub async fn get_task_by_id(id : web::Path<i32>) ->  Result<Json<dto::Task>> {
        println!("Should now get a task with id {}", id.into_inner());
        Ok(Json(Task:rand()))
    }

```

<b>What I would love to see:</b>

1. A "default_ok" status code which could be used with the IntoResponse derive.   
    Why? We often use the same Response dto for all C/R/U/D ops. 
         Thus mapping the status code 1:1 to my dto doesn't make sense for how I work.

2. Ideally there would be a blanket 
   1. `impl<T:ToSchema/ToResponse> IntoResponses for Result<T, E>` or  
   2. `impl<T:ToSchema/ToResponse, E:ToSchema/ToResponse> IntoResponses for Result<T, E>`  
   The 2nd would ofc be much more correct than the first. However would require that every error we return implements ToResponse.  
   This would be annoying as for us at least, most errors are from upstream crates, thus making it impossible for us to implement the ToResponse directly.
   This would be required as the user of utoipa can not make these implementations themselves.

3. For actix_extras there should e.g. be a `impl IntoResponses/ToResponse for Json<T>` as this can not be implemented by the user. 
   
4. IDK if this is possible - but it would also be nice if everything which implements `ToResponse`  
   could be used with the responses attribute. 
   They would then use the default_ok status code for the operation.

5. Until I started looking into the topic specifically I did not understand the usage of the ToResponse and that 
   with actix_extras I can automatically infer the response
   - This Could be me being slow, or maybe the Usage docs being hard to follow.




<details>
  <summary><b>How I might start using this now</b></summary>

I might implement my own Result type like: 
```rs
pub enum GetResult<Ok> {
    Ok(actix_web::web::Json<Ok>),
    Err(actix_web::Error),
} 


impl<Ok:Serialize> Responder for GetResult<Ok> {
    type Body =  EitherBody<<actix_web::web::Json<Ok> as Responder>::Body>;

    fn respond_to(self, req: &actix_web::HttpRequest) -> actix_web::HttpResponse<Self::Body> {
        match self {
            GetResult::Ok(t) => t.respond_to(req).map_into_left_body(),
            GetResult::Err(e) => e.error_response().map_into_right_body(),
        }
    }
}

impl<Ok: for <'a> ToResponse<'a>> IntoResponses for GetResult<Ok> {
    fn responses() -> std::collections::BTreeMap<String, utoipa::openapi::RefOr<utoipa::openapi::response::Response>> {
        let mut map = std::collections::BTreeMap::new();
        map.insert("200".to_string(), Ok::response().1);
        map
    }
}

/// ... PostResult
```
Don't love it as I'll have to into() all my responses this way, but not sure if there is a different idea.


Or possibly might add some post processing to the generated OpenAPI  
which maps all the OK responses to the correct counterpart, and only require a single custom Result


</details>

**Comment by Tortoaster** [comment]
I would also love to have an `IntoResponses` implementation for `Result<T, E>`, but explicitly with both `T` *and* `E` implementing `IntoResponses`. While I agree most errors would probably not be handled any different than others, I like to have a separate, smaller `ApiError` type with limited details as to what exactly went wrong within some dependency, but just enough detail to explain the gist of the problem to the user. This allows me to exhaustively handle cases like "you forgot to do x" on the client side, so that I can properly translate the error messages for example. For use cases where errors are less relevant, a simple `ToResponse` implementation for the error type saying something went wrong would do the trick.

To make this implementation usable by all relevant types, I think there should also be an `impl<'r, T: ToResponse<'r>> IntoResponses for T`. This is a breaking change, because types that implement both `ToResponse` and `IntoResponses` would have conflicting implementations, but I think it's rare to implement both of these for the same type and generate different responses in each of those implementations. Any type that can be turned into one response can trivially be turned into a map of responses, so most users would probably only have to delete their separate `IntoResponses` implementation.

**Comment by titaniumtraveler** [comment]
> How do you currently use the responses with `utoipa` lib?

I use a shared `IntoResponses` impl for an enum with common errors (like Database, authentication and similar errors) and an `Other` variant that is a generic over the route-specific error.

> Second question is how would you like the responses work, or is there any aspect you would like to change in order to them to be more user friendly?

Based on that usage it would be nice to have something like a `#[into_responses]` attribute to make nesting inside the `#[derive]` macro easier.

**Comment by Erik-Sovereign** [comment]
First of all: Thanks for the crate!

We use it with utoipa-axum like this:

rest.rs
```rust
#[utoipa::path(get, path = ROUTE_SINGLE_ROLE,
	params(
		("rolename" = String, description = "Role name"),
	),
	responses(
		(status = OK, body = RestRole, description = "Role with description", example = json!("INSTALLER")),
		(status = NOT_FOUND, description = "Role does not exist"),
	)
)]
async fn get_role(Path(rolename): Path<String>, State(repo): State<UsersRolesRepository>) -> crate::error::Result<Json<RestRole>> {
	let role = repo.role_by_name(&rolename).await?.ok_or(Error::NotFound)?;

	Ok(Json(RestRole { name: role.name }))
}
```

So we use a result, that maps to the correct Json Response, on an "OK" response. We map errors, that can occur, in a seperate file to response like this:

error.rs
```rust
#[derive(Debug, thiserror::Error)]
pub enum Error {
	#[error("SHA-crypt error")]
	ShaCrypt,
	#[error("resource not found")]
	NotFound,
...
}

impl IntoResponse for Error {
	fn into_response(self) -> Response {
		match self {
			Self::Sqlx(_) | Self::DbUnexpected | Self::TaskJoin(_) | Self::ShaCrypt => {
				error!("Internal server error: {}", self);
				StatusCode::INTERNAL_SERVER_ERROR.into_response()
			}

			Self::NotFound => StatusCode::NOT_FOUND.into_response(),
...
}
```

I think, this workflow is okay, but what I would like is, if I wouldn't have to write
`(status = OK, body = RestRole` and `crate::error::Result<Json<RestRole>>` but that I could somehow but the positive Result of an endpoint as the body of a response. Because right now those two could get out of sync and I always have to check manually that they don't.

**Comment by vaytea** [comment]
I'm using Rocket, I have a general response struct which implements `Responder` trait, and all my APIs return a the struct.
I would like to instead of having something like: 
```rust
#[utoipa::path(
    responses(
        (
            status = 200,
            description = "Editable users retrieved successfully",
            body = APIResponse,
            example = json!({ "message": Message::Successful, "result": {} })
        )
    ),
    security(
        ("api_key" = [])
    )
)]
#[get("/")]
async fn get_users(admin: &User, _access: UserPageAccess, db: Db) -> APIResponse {
    ...
}
```
use the macro like:
``` rust
#[utoipa::path(
    responses(
        APIResponse::SUCCESS.with_description("Editable users retrieved successfully"),        
    ),
    security(
        ("api_key" = [])
    )
)]
#[get("/")]
async fn get_users(admin: &User, _access: UserPageAccess, db: Db) -> APIResponse {
    ...
}
```
I think that would be possible through implementing a trait like `ToExample`

**Comment by jonasehrlich** [comment]
Hi, first of all thank you for this crate.

What I would love to see is less boilerplate when defining API endpoints. I am using `axum` and the `utoipa-axum` crate. For all of my endpoints I want to use a common status response schema (at least for errors!).

```rust
#[derive(Serialize, Debug, ToSchema, Deserialize)]
#[serde(rename_all = "camelCase")]
/// Basic serializable API status response
pub struct StatusResponse {
    /// HTTP status code
    status: u16,
    /// Canonical reason for the status
    reason: String,
    /// Message with details about the status
    message: String,
}
```

Currently every endpoint is specifying the same error responses again and again:

```rust
#[utoipa::path(
    get,
    path = "/fruits",
    params(GetFruitsQueryParams),
    responses(
        (status = http::StatusCode::OK, description = "Ok", body = GetFruitsResponse),
        (status = http::StatusCode::BAD_REQUEST, description = "Bad Request", body = api::StatusResponse),
        (status = http::StatusCode::NOT_FOUND, description = "Not Found", body = api::StatusResponse),
        (status = http::StatusCode::INTERNAL_SERVER_ERROR, description = "Internal Server Error", body = api::StatusResponse)
    )
)]
```

What would be nice (although I am not sure if it is possible with the capabilities of macros!) would be support for a generic enum that implements `IntoResponses`:

```rust
enum EndpointResponses<T: ToResponse> {
    Ok(T),
    NotFound,
    BadRequest,
    InternalServerError,
}

impl<T: ToResponse> IntoResponses for EndpointResponses<T> {
    fn responses() -> BTreeMap<String, openapi::RefOr<openapi::Response>> {
        openapi::ResponsesBuilder::new()
            .response("200", T::response().into())
            .response(
                "400",
                openapi_response_from_http_status(http::StatusCode::BAD_REQUEST),
            )
            .response(
                "404",
                openapi_response_from_http_status(http::StatusCode::NOT_FOUND), 
            )
            .response(
                "500",
                openapi_response_from_http_status(http::StatusCode::INTERNAL_SERVER_ERROR), //
            )
            .build()
            .into()
    }
}
```

That way, the responses could become just a 

```rust
responses(EndpointResponses<GetFruitsResponse>)
```

And the different responses could be maintained in one place.

Something else I noticed is, that when deriving `ToResponse` on `StatusResponse` and using the shorthand `response = ...` syntax the content type is missing from the schema, except if `response = inline(...)` is used. Am I missing something here?

```rust
#[utoipa::path(
    get,
    path = "/fruits",
    params(GetFruitsQueryParams),
    responses(
        (status = http::StatusCode::BAD_REQUEST, response = api::StatusResponse),
    )
)]
```
